### PR TITLE
Remove empty parts after they were pruned by TTL [2]

### LINKED ISSUE #5491: Merges on a table with TTL leave empty partitions
It's possible that after a merge on a table with TTL there will be no more rows in some partitions. It makes sense to drop those partitions.
```
drop table if exists ttl_00933_1;
create table ttl_00933_1 (d Date, a Int) engine = MergeTree order by a partition by toYear(d) ttl d + interval 1 day;
insert into ttl_00933_1 values (toDate('2000-10-10'), 1);
insert into ttl_00933_1 values (toDate('2100-10-10'), 2);
optimize table ttl_00933_1 final;
select partition, name, active, rows from system.parts where table = 'ttl_00933_1' and active

SELECT
    partition,
    name,
    active,
    rows
FROM system.parts
WHERE (table = 'ttl_00933_1') AND active

┌─partition─┬─name───────┬─active─┬─rows─┐
│ 2000      │ 2000_1_1_1 │      1 │    0 │
│ 2100      │ 2100_2_2_0 │      1 │    1 │
└───────────┴────────────┴────────┴──────┘
```

### LINKED ISSUE #15180: TTL_ONLY_DROP_PARTS does not drop empty parts
I created table with following structure and inserted some test data. 
Empty parts(rows = 0) still exists even if all rows from the table were deleted over TTL.

create table  temp_ttl_partition_test
 (
        rec_date DateTime,
        user_id String
)
engine = MergeTree
partition by formatDateTime(rec_date, '%Y%m%d%H')
ORDER BY user_id
TTL rec_date + interval 10 minute
SETTINGS ttl_only_drop_parts=1;

CH version = 20.5.2.7

Expected behavior: TTL_ONLY_DROP_PARTS must drop empty parts

bug

### LINKED ISSUE #16629: Why not remove outdated parts permanently after TTL merge?
##  Description 

I find some empty parts after background merge. These parts are old enough to be deleted permanently. Why does it still exists?

image

image

### Question1: Does it work as expect? 

I have read the source code about background merge logic, but not found logic to delete it.

### Question2: Can it cause "too many parts" problem?

Maybe, it can exceeds the parts amount limitation `max_parts_in_total`.

Now, i can avoid this problem by the following two steps:

* Wait for background merge or trigger merge manually by "OPTIMIZE xxx"  to ensure that only one part for one outdated partition.
* Drop outdated partition by "ALTER TABLE table_name DROP PARTITION XXX".

### Question3: Why not remote it permanently?

Why not do this? Are there any other considerations?

CH version 19.16.20.1.

### PR DESCRIPTION
**Pull request by CurtizJ**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Remove empty parts after they were pruned by TTL, mutation, or collapsing merge algorithm.

Detailed description / Documentation draft:
Based on #15511.

Fixes #16629. Fixes #15180. Fixes #5491.

### CODE REVIEW

**Comment by CurtizJ** [comment]
This PR contains also #16875 and should be merged after it.

File: src/Storages/MergeTree/MergeTreeData.cpp [hunk 1]
```diff
@@ -1215,6 +1215,19 @@ void MergeTreeData::clearOldWriteAheadLogs()
...
 }
 
+void MergeTreeData::clearEmptyParts()
+{
+    auto parts = getDataPartsVector();
+    for (const auto & part : parts)
+    {
+        if (part->rows_count == 0)
+        {
+            ASTPtr literal = std::make_shared<ASTLiteral>(part->name);
```
Comment (alesapin) [inline src/Storages/MergeTree/MergeTreeData.cpp:1225]: quite ugly, better to have separate method for AST, but ok.

File: src/Storages/MergeTree/ReplicatedMergeTreeCleanupThread.cpp [hunk 2]
```diff
@@ -72,6 +72,7 @@ void ReplicatedMergeTreeCleanupThread::iterate()
         clearOldLogs();
         clearOldBlocks();
         clearOldMutations();
+        storage.clearEmptyParts();
```
Comment (alesapin) [inline src/Storages/MergeTree/ReplicatedMergeTreeCleanupThread.cpp:75]: What if multiple replicas will assign `DROP_RANGE` for the same parts almost at the same time. Will it work well? Maybe we will have some exceptions in logs?

**Comment by alesapin** [comment]
Discussed in telegram -- need tiny fixes.

**Review by alesapin — APPROVED** [review]
LGTM

**Pull request merged**